src/requests/create-signature.ts

```diff
@@ -0,0 +1,120 @@
+import { getNormalizedEncodedURI, getNormalizedHeaders } from './utils'
+import { CONTENTFUL_HEADERS, ContentfulHeader } from './constants'
+
+const hash = (normalizedCanonicalRequest: NormalizedCanonicalRequest, secret: string) => {
+  const stringifiedHeaders = normalizedCanonicalRequest
+    .headers!.map(([key, value]) => `${key}:${value}`)
+    .join(';')
+
+  const stringifiedRequest = [
+    normalizedCanonicalRequest.method,
+    normalizedCanonicalRequest.path,
+    stringifiedHeaders,
+    normalizedCanonicalRequest.body,
+  ].join('\n')
+
+  const hmac = crypto.createHmac('sha256', secret)
+
+  hmac.update(stringifiedRequest)
+
+  return hmac.digest('hex')
+}
+
+const removeExistingContentfulHeaders = (headers: NormalizedHeaders) => {
+  return headers.filter(([key]) => !CONTENTFUL_HEADERS.includes(key))
+}
+
+const enrichNormalizedHeadersWithMetadata = (headers: NormalizedHeaders, timestamp: number) => {
+  const result = removeExistingContentfulHeaders(headers)
+  const joinedSignedHeaders = result
+    .map(([key]) => key)
+    // We always sign timestamp and signed headers
+    .concat(ContentfulHeader.Timestamp, ContentfulHeader.SignedHeaders)
+    .join(',')
+
+  result.push([ContentfulHeader.Timestamp, timestamp.toString()])
+  result.push([ContentfulHeader.SignedHeaders, joinedSignedHeaders])
+
+  return result
+}
+
+/**
+ * Given a secret, a canonical request and a timestamp, generates a signature.
+ * It can be used to verify canonical requests to assess authenticity of the
+ * sender and integrity of the payload.
+ *
+ * ~~~
+ * const {createSignature} = require('contentful-node-apps-toolkit')
+ * const {pick} = require('lodash')
+ * const {server} = require('./imaginary-server')
+ *
+ * const SECRET = process.env.SECRET
+ *
+ * server.post('/api/my-resources', (req, res) => {
+ *   const incomingSignature = req.headers['x-contentful-signature']
+ *   const incomingTimestamp = req.headers['x-contentful-timestamp']
+ *   const incomingSignedHeaders = req.headers['x-contentful-signed-headers']
+ *   const now = Date.now()
+ *
+ *   if (!incomingSignature) {
+ *     res.send(400, 'Missing signature')
+ *   }
+ *
+ *   if (now - incomingTimestamp > 1000) {
+ *     res.send(408, 'Request too old')
+ *   }
+ *
+ *   const signedHeaders = incomingSignedHeaders.split(',')
+ *
+ *   const computedSignature = createSignature(
+ *     SECRET,
+ *     {
+ *       method: req.method,
+ *       path: req.url,
+ *       headers: pick(req.headers, signedHeaders),
+ *       body: JSON.stringify(req.body)
+ *     },
+ *     incomingTimestamp
+ *   )
+ *
+ *   if (computedSignature !== incomingSignature) {
+ *      res.send(403, 'Invalid signature')
+ *   }
+ *
+ *   // rest of the code
+ * })
+ *
+ * ~~~
+ * @category Requests
+ */
+export const createSignature = (
+  rawSecret: Secret,
+  rawCanonicalRequest: CanonicalRequest,
+  rawTimestamp: Timestamp
```

Comment:
I wouldn't require a timestamp, just use the current time (keep the argument for testing)